### PR TITLE
8234446: Post-CMS workgroup hierarchy cleanup

### PR DESCRIPTION
**Pull request by albertnetymk**
Collapse workgroup hierarchy and some general cleanup.

Tested; hotspot_gc

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8234446](https://bugs.openjdk.java.net/browse/JDK-8234446): Post-CMS workgroup hierarchy cleanup

### Reviewers
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Roman Kennke](https://openjdk.java.net/census#rkennke) (@rkennke - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3663/head:pull/3663` \
`$ git checkout pull/3663`

Update a local copy of the PR: \
`$ git checkout pull/3663` \
`$ git pull https://git.openjdk.java.net/jdk pull/3663/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3663`

View PR using the GUI difftool: \
`$ git pr show -t 3663`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3663.diff">https://git.openjdk.java.net/jdk/pull/3663.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/shared/workgroup.cpp [hunk 1]
```diff
@@ -198,20 +127,77 @@ class GangTaskDispatcher : public CHeapObj<mtGC> {
...
 
-WorkGang::WorkGang(const char* name,
-                   uint  workers,
-                   bool  are_GC_task_threads,
-                   bool  are_ConcurrentGC_threads) :
-    AbstractWorkGang(name, workers, are_GC_task_threads, are_ConcurrentGC_threads),
-    _dispatcher(new GangTaskDispatcher())
-{ }
+WorkGang::WorkGang(const char* name, uint workers, bool are_GC_task_threads, bool are_ConcurrentGC_threads) :
+      _workers(NULL),
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/workgroup.cpp:133]: Pre-existing: the indentation is completely arbitrary here. Please use something like 2 spaces.

File: src/hotspot/share/gc/shared/workgroup.hpp [hunk 2]
```diff
@@ -87,10 +82,10 @@ struct WorkData {
 // The work gang is the collection of workers to execute tasks.
 // The number of workers run for a task is "_active_workers"
 // while "_total_workers" is the number of available of workers.
-class AbstractWorkGang : public CHeapObj<mtInternal> {
- protected:
+class WorkGang : public CHeapObj<mtInternal> {
+ private:
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/workgroup.hpp:86]: The `private` can be removed.

File: src/hotspot/share/gc/shared/workerManager.hpp [hunk 3]
```diff
@@ -50,25 +50,22 @@ class WorkerManager : public AllStatic {
   // create all the worker at start should considered a problem so exit.
   // If initializing = false, there are already some number of worker
   // threads and a failure would not be optimal but should not be fatal.
-  template <class WorkerType>
-  static uint add_workers (WorkerType* holder,
+  static uint add_workers (WorkGang* holder,
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/workerManager.hpp:53]: s/holder/workers

File: src/hotspot/share/gc/shared/workerManager.hpp [hunk 4]
```diff
@@ -50,25 +50,22 @@ class WorkerManager : public AllStatic {
...
                            uint active_workers,
                            uint total_workers,
                            uint created_workers,
                            os::ThreadType worker_type,
                            bool initializing);
 
   // Log (at trace level) a change in the number of created workers.
-  template <class WorkerType>
-  static void log_worker_creation(WorkerType* holder,
+  static void log_worker_creation(WorkGang* holder,
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/workerManager.hpp:61]: s/holder/workers

**Review by tschatzl — CHANGES_REQUESTED** [review]
Mostly further cleanup suggestions.

**Comment by albertnetymk** [comment]
There is some preexisting inconsistency regarding the indentation of fields initializers in constructors, e.g. `SubTasksDone` uses 2-space, while `WorkGang` uses 4-space. In order not to make the review process unnecessarily harder, I have kept the original indentation, despite the inconsistency.

**Review by tschatzl — APPROVED** [review]

**Review by rkennke — APPROVED** [review]
Looks good to me. Thank you!

**Comment by albertnetymk** [comment]
Thanks for the review.

/integrate

**Pull request closed without merging**